File: ResourceManager/ViewModels/Passwords/PasswordContentViewModel.cs
using Microsoft.Extensions.DependencyInjection;
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Commands;
using NetCore.Mvvm.Controls;
using ResourceManager.Configuration;
using ResourceManager.Helpers;
using ResourceManager.Models;
using ResourceManager.Services.Abstractions;
using System;
using System.Linq;
using System.Windows.Input;

namespace ResourceManager.ViewModels
{
    public class PasswordContentViewModel : PageContentWithListViewModel<Password>
    {
        private readonly IDataServiceManager _dataServiceManager;
        public PasswordContentViewModel(IDataServiceManager dataServiceManager)
            : base(dataServiceManager.PasswordDataService)
        {
            _dataServiceManager = dataServiceManager;
            MenuTitle = "Password";
            MenuIcon = AssetManager.Instance.GetImage("Passwords.png");
            Title = "Password Manager";
            PageColor = Constants.PasswordPageColor;

            _add_permission_name = Constants.ACTION_ADD_PASSWORD;
            _edit_permission_name = Constants.ACTION_EDIT_PASSWORD;
            _delete_permission_name = Constants.ACTION_DEL_PASSWORD;

            TagsCommand = new RelayCommand<Password>(OnTags);

            OnRefresh();
        }

        private async void OnTags(Password pass)
        {
            var vm = App.ServiceProvider.GetRequiredService<TagsPopupViewModel>();
            vm.Caption = string.Format("Manage tag for {0}", pass.GetCaption());
            var tags = await _dataServiceManager.TagDataService.GetTagObjectForPasswordID(pass.ID);
            vm.Tags = tags.ToList();
            var popup = ShowPopup(vm);
            if (popup?.Result == PopupResult.OK)
            {
                await _dataServiceManager.PasswordDataService.UpdateTags(pass, vm.Tags);
                OnRefresh();
            }
        }

        protected override void OnEdit(Password password)
        {
            var vm = App.ServiceProvider.GetRequiredService<PasswordPopupViewModel>();
            vm.Caption = "Edit Password";
            vm.SetPassword(password);
            var popup = ShowPopup(vm);
            if (popup?.Result == PopupResult.OK)
            {
                OnRefresh();
            }
        }

        protected override void OnAdd()
        {
            var vm = App.ServiceProvider.GetRequiredService<PasswordPopupViewModel>();
            vm.Caption = "Add Password";
            var popup = ShowPopup(vm);
            if (popup?.Result == PopupResult.OK)
            {
                OnRefresh();
            }
        }
        protected override void OnFilter()
        {
            Filter(FilterColumns.Password, "FIlter Password");
        }

        public ICommand TagsCommand { get; }
    }
}

File: ResourceManager/ViewModels/Passwords/PasswordContentViewModel.vb
Imports VBNetCore.Mvvm.Abstractions
Imports VBNetCore.Mvvm.Commands
Imports VBNetCore.Mvvm.Controls
Imports ResourceManager.Configuration
Imports ResourceManager.Helpers
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions
Imports System.Linq
Imports System.Windows.Input
Imports Microsoft.Extensions.DependencyInjection

Namespace ResourceManager.ViewModels
    Public Class PasswordContentViewModel
        Inherits PageContentWithListViewModel(Of Password)
        Private ReadOnly _dataServiceManager As IDataServiceManager
        Public Sub New(ByVal dataServiceManager As IDataServiceManager)
            MyBase.New(dataServiceManager.PasswordDataService)
            _dataServiceManager = dataServiceManager
            MenuTitle = "Password"
            MenuIcon = AssetManager.Instance.GetImage("Passwords.png")
            Title = "Password Manager"
            PageColor = PasswordPageColor

            _add_permission_name = ACTION_ADD_PASSWORD
            _edit_permission_name = ACTION_EDIT_PASSWORD
            _delete_permission_name = ACTION_DEL_PASSWORD

            TagsCommand = New RelayCommand(Of Password)(AddressOf OnTags)

            MyBase.OnRefresh()
        End Sub

        Private Async Sub OnTags(ByVal pass As Password)
            Dim vm = Application.ServiceProvider.GetRequiredService(Of TagsPopupViewModel)()
            vm.Caption = String.Format("Manage tag for {0}", pass.GetCaption())
            Dim tags = Await _dataServiceManager.TagDataService.GetTagObjectForPasswordID(pass.ID)
            vm.Tags = tags.ToList()
            Dim popup = ShowPopup(vm)
            If popup.Result = PopupResult.OK Then
                Await _dataServiceManager.PasswordDataService.UpdateTags(pass, vm.Tags)
                MyBase.OnRefresh()
            End If
        End Sub

        Protected Overrides Sub OnEdit(ByVal password As Password)
            Dim vm = Application.ServiceProvider.GetRequiredService(Of PasswordPopupViewModel)()
            vm.Caption = "Edit Password"
            vm.SetPassword(password)
            Dim popup = ShowPopup(vm)
            If popup.Result = PopupResult.OK Then
                MyBase.OnRefresh()
            End If
        End Sub

        Protected Overrides Sub OnAdd()
            Dim vm = Application.ServiceProvider.GetRequiredService(Of PasswordPopupViewModel)()
            vm.Caption = "Add Password"
            Dim popup = ShowPopup(vm)
            If popup.Result = PopupResult.OK Then
                MyBase.OnRefresh()
            End If
        End Sub
        Protected Overrides Sub OnFilter()
            MyBase.Filter(FilterColumns.Password, "FIlter Password")
        End Sub

        Public ReadOnly Property TagsCommand As ICommand
    End Class
End Namespace
